File: easy/Best Time to Buy and Sell Stock/maxProfit.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function (prices) {
  if (prices.length < 1) {
    return 0;
  }

  let broughtFor = prices[0];
  let maxProfit = 0;

  for (let i = 1; i < prices.length; ++i) {
    if (prices[i] < broughtFor) {
      broughtFor = prices[i];
    } else {
      maxProfit = Math.max(maxProfit, prices[i] - broughtFor);
    }
  }

  return maxProfit;
};

const testMaxProfit = function (prices, expected) {
  const result = maxProfit(prices);
  if (result == expected) {
    console.log("Test Case Passed!");
  } else {
    console.log(
      `Test Case Failed: prices: ${prices} expected: ${expected} result: ${result}`
    );
  }
};

testMaxProfit([7, 1, 5, 3, 6, 4], 5);
testMaxProfit([7, 6, 4, 3, 1], 0);
testMaxProfit([1], 0);

File: second_pass/3_sliding_window/1_buy_stock/max_profit.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if len(prices) == 0:
            return 0

        n = len(prices)
        brought_for = prices[0]
        max_profit = 0

        for i in range(1, n):
            if prices[i] < brought_for:
                brought_for = prices[i]
            else:
                max_profit = max(max_profit, prices[i] - brought_for)

        return max_profit

    def testMaxProfit(self, prices, expected):
        actual = self.maxProfit(prices)
        if actual == expected:
            print("Test Case Passed!")
        else:
            print(f"Test Case Failed!")


Solution().testMaxProfit([7, 1, 5, 3, 6, 4], 5)
Solution().testMaxProfit([7, 6, 4, 3, 1], 0)
